trunk/src/modules/upscale.F90

```diff
@@ -110,68 +160,97 @@ subroutine load_vis(V,Fp,Etype)
       end do
 !
       read(nin,*) V%NR_ELEM
-      allocate(V%ELEM(1:V%NR_ELEM, 0:8), STAT=istat)
+      allocate(V%ELEM(1:V%NR_ELEM, 0:27), STAT=istat)
       V%ELEM = 0
       ierr = ierr + istat
 !
-      do i=1,V%NR_ELEM
-         select case (Etype)
-         case(TETR,MDLN)
-            read(nin,*) &
-               V%ELEM(i,0), &
-               V%ELEM(i,1), V%ELEM(i,2), V%ELEM(i,3), V%ELEM(i,4)
-         case(PRIS,MDLP)
-            read(nin,*) &
-               V%ELEM(i,0), &
-               V%ELEM(i,1), V%ELEM(i,2), V%ELEM(i,3),&
-               V%ELEM(i,4), V%ELEM(i,5), V%ELEM(i,6)
-         case(BRIC,MDLB)
-            read(nin,*) &
-               V%ELEM(i,0), &
-               V%ELEM(i,1), V%ELEM(i,2), V%ELEM(i,3), V%ELEM(i,4), &
-               V%ELEM(i,5), V%ELEM(i,6), V%ELEM(i,7), V%ELEM(i,8)
-         end select
-      end do
+!
+      if (SECOND_ORDER_VIS) then
+         read(nin,*) V%ELEM(1,0)
+         do i=1,V%NR_VERT
+            read(nin,*) V%ELEM(1,i)
+         enddo
+      else
+         do i=1,V%NR_ELEM
+            select case (Etype)
+            case(TETR,MDLN)
+               read(nin,*) &
+                  V%ELEM(i,0), &
+                  V%ELEM(i,1), V%ELEM(i,2), V%ELEM(i,3), V%ELEM(i,4)
+            case(PRIS,MDLP)
+               read(nin,*) &
+                  V%ELEM(i,0), &
+                  V%ELEM(i,1), V%ELEM(i,2), V%ELEM(i,3),&
+                  V%ELEM(i,4), V%ELEM(i,5), V%ELEM(i,6)
+            case(BRIC,MDLB)
+               read(nin,*) &
+                  V%ELEM(i,0), &
+                  V%ELEM(i,1), V%ELEM(i,2), V%ELEM(i,3), V%ELEM(i,4), &
+                  V%ELEM(i,5), V%ELEM(i,6), V%ELEM(i,7), V%ELEM(i,8)
+            end select
+         enddo
+      end if
 !
       close(nin)
       ierr = ierr + istat
       if (ierr.ne.0) then
          write(*,*) 'module upscale::Something wrong', ierr
       end if
-   end subroutine load_vis
 !
+   end subroutine load_vis
+
+
+
 !> @brief return point coordinates of a point associated with a vis object
 !> @param[in]  V
 !> @param[in]  Idx
 !> @param[out] Pt
 !> @date Feb 2023
    subroutine get_vis_point(V,Idx, Pt)
+!
+      implicit none
+!
       type(vis), intent(in) :: V
       integer  , intent(in) :: Idx
       real(8)  , intent(out):: Pt(3)
+!
       Pt(1:3) = V%VERTC(Idx,1:3)
-   end subroutine get_vis_point
 !
+   end subroutine get_vis_point
+
+
+
 !> @brief return list of vertices for vis object
 !> @param[in]  V
 !> @param[in]  Idx
 !> @param[in]  Ioffs
 !> @param[out] Iverl
 !> @date Feb 2023
    subroutine get_vis_elem(V,Idx,Ioffs, Iverl)
+!
+      implicit none
+!
       type(vis), intent(in)  :: V
       integer  , intent(in)  :: Idx, Ioffs
-      integer  , intent(out) :: Iverl(8)
-      Iverl(1:8) = V%ELEM(Idx,1:8) + Ioffs
-   end subroutine get_vis_elem
+      integer  , intent(out) :: Iverl(27)
+
+      Iverl(1:27) = V%ELEM(Idx,1:27) + Ioffs
```

Comment:
I think it'd be good to distinguish between the linear and second-order case, either by using a parameter for the array size, or by using two different `get_vis_elem` functions depending on the case. Otherwise, the linear routines become less efficient.

Reply:
I don't like the idea of adding another parameter to specify the size, it just makes it harder to use. Does it really make sense to make a separate version for second order elements? Then you have to select which function to use based on `SECOND_ORDER_VIS`. We could only copy the relevant entries for clarity but it's really not an efficiency concern; the cost of copying 27 elements is dwarfed in magnitude by the cost `nodcor` and `solelm` (which are defining arrays with sizes defined by the maximum order and doing a lot of actual computation)

Reply:
Doesn't have to be a separate function or parameterized, but here I'd also use
```suggestion
      if (SECOND_ORDER_VIS) then
         Iverl(1:27) = V%ELEM(Idx,1:27) + Ioffs
      else
         Iverl(1:8) = V%ELEM(Idx,1:8) + Ioffs
      endif
```
then it is clear that only 8 are used in the linear case.

Reply:
Sounds good!